pkgs/build-support/singularity-tools/default.nix

```diff
@@ -82,7 +83,17 @@ rec {
               util-linux
             ];
             strictDeps = true;
-            layerClosure = writeClosure contents;
+            layerClosure = writeClosure (
+              [
+                bashInteractive
+                # runScript references
+                # TODO(@ShamrockLee):
+                # Implement fine-grain reference parsing with writeStringReferencesToFile
+                # once we make writers.makeScriptWriter expose the .text attribute.
+                (writeDirectReferencesToFile runScriptFile)
+              ]
```

Comment:
I didn't understand the comment or how this bit relates to the PR description. Why two layers of derivations for extracting the references?

Reply:
Here's what I would do instead if we have `runScriptFile.text`:

```nix
runCommand {
  layerClosure = writeClosure ([
    bashInteractive
  ] ++ lib.splitString "\n" (writeStringReferencesToFile runScriptFile.text).text)
  runScriptText = runScriptFile.text;
  passAsFile = [ "runScriptFile" ]
} ''
  ...
  mv "$runScriptText" singularity.d/runscript
''
```

This way, we don't need to instantiate and build the intermediate derivation `runScriptFile` while still getting it into the container. We don't even have to write the `runScript` content twice.

Reply:
We could achieve this by switching back to `writeShellScript` to construct `runScriptFile`, but I prefer to fix the `writer.makeScriptWriter` to keep using the unified interface of `writers.writeBash`.

Reply:
Hmm what's `(writeStringReferencesToFile runScriptFile.text).text`

Reply:
> Hmm what's `(writeStringReferencesToFile ...).text`

If `builtins.getContext` is available, writeStringReferencesToFile will use it to get the reference paths, compose to a string, append the string context, and write the string with `writeText`. `writeText` exposes the input string through the `text` attribute.

Reply:
It turns out to be hard to make `writers.makeScriptWriter` expose the `text` attribute, as it works around Darwin's prohibition of script-as-shebang-interpreter by determining at build time if the input `interpreter` is a script and constructing a custom shebang.

I'll switch it back to `writeScript` then.

Reply:
> If builtins.getContext is available, writeStringReferencesToFile will use it to get the reference paths,

If I understood you correctly, this means "all store paths from `inputDrvs` (=direct dependencies, not transitive) referenced by the environment or the builder script (even those paths that are not referenced in the output)"? If so, why that? I'm missing the high level picture

> string context

Oy vey, I'd rather not rely on string contexts. E.g. TVix will not implement them

Reply:
It's `runInLinuxVM`'s fault again. It doesn't respect the `passAsFile` specification of the input attribute (#16742). Building and copying `runScriptFile` seems to be the most sensible implementation.

Reply:
I just pushed the revised version. Please take a look.

Reply:
> > string context
>
> Oy vey, I'd rather not rely on string contexts. E.g. TVix will not implement them

Thanks for letting me know. The new implementation checks whether `builtins.getContext` is supported before choosing the referencing approach.